RIOT-OS/RIOT
# congure: initial import of a congestion control framework

### Contribution description
This provides _CongURE_ (*Cong*estion Control *U*tilizing *R*eusable *E*lements, pronounced like "conjure"), a framework to implement congestion control and use it in a reusable manner with a number of protocols.

The original motivation was to provide congestion control with [6LoWPAN selective fragment recovery](https://doc.riot-os.org/group__net__gnrc__sixlowpan__frag__sfr.html) and allow for the congestion control mechanism to be interchangeable. However, now that there is a framework with multiple well tested implementations of congestion control, it can also be used for other protocols, such as TCP or QUIC, as I kept the units for the congestion window up to the user (TCP and QUIC, e.g., uses bytes, SFR, e.g., uses number of fragments). It is also designed as such that multiple instances of the same implementation can run on the same image (e.g. for when TCP and SFR want to use the same CC). In that case, I can imagine a lot of ROM can be saved, since there is only one CC code for two layers.

A test framework for the CongURE framework is also provided in form of shell commands to trigger the API functions of CongURE, as well as tests for said framework (a framework test framework test if you want :wink:). If preferred, I can move those to a separate PR.

As to why this is in `sys/` and not `sys/net/`: There are no networking components to this framework, it just provides the functions to calculate the size of a congestion window. `seq` e.g. is also in `sys/`, though sequence number arithmetic mostly shows up in networking. Not sure this could help with local queue congestion problems, but when thinking out of the box, one might think as such :-P.



### Testing procedure
A test framework including tests for it is provided. [`riotctrl`](https://github.com/RIOT-OS/riotctrl/) (at least v0.2.2) is required for them to run. If installed,

```
make -j -C tests/congure_test flash test
```

should succeed.



### Issues/PRs references
Follow-up PRs for actual implementations of congestion control mechanisms will follow, as well as integration PRs for GNRC's 6LoWPAN SFR and maybe TCP as well :-). When merged, I will adapt the test framework to use the turo framework introduced in #15950, which was the original idea to begin with (which is why it is already printing JSON strings ;-)).
![PR dependency graph](https://page.mi.fu-berlin.de/mlenders/congure.svg)
